### ISSUE TITLE
"index was out of range"...

### ISSUE THREAD

**Issue by IAmOrion** [ScottFerg56/GFXFontEditor#1]
Hi, firstly, great little program you've written here, very useful!

I do have an issue though.  When the .h file is large - 558kb - the font being 96pt the program "breaks" - albeit gracefully, with the following error.

<img width="425" alt="Screenshot 2024-07-19 at 02 15 50" src="https://github.com/user-attachments/assets/2b134d71-0c4e-4314-9d22-bb60a565906f">

[Audi96pt7b.h.zip](https://github.com/user-attachments/files/16300785/Audi96pt7b.h.zip)

(It worked for sizes 12, 24, 48 without issue)

Kind regards

**Comment by ScottFerg56** [comment]
The Audi96pt7b.h font file is actually invalid. The GFXglyph bitmapOffset field supports 16-bit integer offsets, thus the header file format only supports 64K of 'Bitmaps' bytes, whereas this file contains 89536 bytes. So you'll see that, starting at line 7532, the offset wraps around to a lower, invalid number. This causes an error in GFXFontEditor. What was the source of this invalid file?

I'll leave this issue open as a note to add a more descriptive error message for this specific invalid file case.

Thanks!

**Comment by IAmOrion** [comment]
> The Audi96pt7b.h font file is actually invalid. The GFXglyph bitmapOffset field supports 16-bit integer offsets, thus the header file format only supports 64K of 'Bitmaps' bytes, whereas this file contains 89536 bytes. So you'll see that, starting at line 7532, the offset wraps around to a lower, invalid number. This causes an error in GFXFontEditor. What was the source of this invalid file?
> 
> I'll leave this issue open as a note to add a more descriptive error message for this specific invalid file case.
> 
> Thanks!

Thanks for the response, that was somewhat over my head ha

The font file is this:

[AudiTypeVF_4.04.ttf.zip](https://github.com/user-attachments/files/16309746/AudiTypeVF_4.04.ttf.zip)

but to save space for my usage, I used FontForge to remove the characters that are not needed to reduce the size of the font file.  Once I had "Detached & Removed" the glyphs / characters not being used, I then exported the results as a new ttf file 

[Audi.ttf.zip](https://github.com/user-attachments/files/16309781/Audi.ttf.zip)

I then used TrueType2GFX website (https://rop.nl/truetype2gfx/) to generate the header file, which I then was editing further using your GFX Font Editor

If you know any better TTF font editors where I can remove unsed characters easily let me know! I found many font editors out there but couldn't see anyways to remove characters to minimise the font before editing further in GFX Font Editor 

For what it's worth, the process **worked** for all sizes up to 82pt - anything bigger than 82pt is when it fails
GFX Font editor was able to open all the other sizes I was testing but 82 seems to be the limit but sounds like that's due to problems in the font/header

**Comment by ScottFerg56** [comment]
Yes, larger font sizes will require more bitmap space and eventually overflow the 64K limit. There are some gaps in the character space in your header file that are being filled with rectangle glyphs where they could just be blank, saving some space. That may be the fault of TrueType2GFX since Audi.ttf shows blank glyphs for the gaps when loaded into FontForge.

You could try saving to BDF format from FontForge and loading that into GFXFontEditor. That conversion is a complicated process with lots of variables you may want to play with. See https://learn.adafruit.com/custom-fonts-for-pyportal-circuitpython-display/conversion. And you will likely still hit the 64K limit for header files. GFXFontEditor will also likely, without announcing an error, produce a file with invalid GFXglyph bitmapOffset field values exceeding 64K.

Let me know how it works out!

**Comment by IAmOrion** [comment]
Thanks for your help and feedback.  I tried FontLab8 yesterday instead, everything worked perfectly (I only needed size 114pt in the end.  Only using the characters 1234567890PRNDSCF - all others were removed and then the glyph/font "flattened".  Ran it through TrueType2GFX and GFX Font Editor and had no issues this time.  

Probably down to my lack of understanding regarding the complexities of fonts and also for whatever reason (again probably my poor understanding/usage of FontForge itself) FontForge was generating junk.  

FontLab8 + GFX Font Editor etc worked perfectly first time!  

I'll close this since I've now achieved what I needed.  Thank you again for both the software and taking the time to respond
